# scriptsForBoltEtAl2021

All the scripts needed to reproduce the NGS figures in Bolt et al. 2021 from raw data.

genomes directory gives the scripts to get the mm10 wt fasta file and the fasta file for the inv2 mutant genome.

The CUT&RUN, CHi-C, and ATACseq preprocessing were done on our galaxyserver. A summary of command lines used are given. Other commands (R scripts, deepTools) which were specific to one experiment type are in these folders as well as the output before modification with illustrator.

The Plots directory show all command lines for pyGenomeTracks plots of Figure 2 and Supplementary Figure 2. It need to be done from the files available on GEO or the results of scripts in the directories cited above. In the Plots/outputs, you can find the genomic figures before modification with illustrator.

In scRNAseq, you can find all command lines and scripts used and the plots used in the figures before modification with illustrator.
